src/saga_action_func.rs

```diff
@@ -135,125 +83,88 @@ where
      * interfaces of its own so there's generally no need to have the specific
      * type.)
      */
-    pub fn new_action(
+    pub fn new_action<UserType, ActionFuncOutput>(
         action_func: ActionFuncType,
         undo_func: UndoFuncType,
-    ) -> Arc<dyn Action<UserType>> {
-        Arc::new(ActionFunc { action_func, undo_func, phantom: PhantomData })
+    ) -> Arc<dyn Action<UserType>>
+    where
+        UserType: SagaType,
+        for<'c> ActionFuncType: ActionFn<
+            'c,
+            UserType,
+            Output = ActionFuncResult<ActionFuncOutput, ActionError>,
+        >,
+        ActionFuncOutput: ActionData,
+        for<'c> UndoFuncType: ActionFn<'c, UserType, Output = UndoResult>,
+    {
+        Arc::new(ActionFunc { action_func, undo_func })
     }
 }
 
 /*
  * TODO-cleanup why can't new_action_noop_undo live in the Action namespace?
  */
 
-async fn undo_noop<UserType: SagaType>(
-    _: ActionContext<UserType>,
-) -> UndoResult {
-    // TODO-log
-    Ok(())
-}
-
 /**
  * Given a function `f`, return an `ActionFunc` that uses `f` as the action and
  * provides a no-op undo function (which does nothing and always succeeds).
  */
-pub fn new_action_noop_undo<
-    UserType,
-    ActionFutType,
-    ActionFuncType,
-    ActionFuncOutput,
->(
+pub fn new_action_noop_undo<UserType, ActionFuncType, ActionFuncOutput>(
     f: ActionFuncType,
 ) -> Arc<dyn Action<UserType>>
 where
     UserType: SagaType,
-    ActionFuncType:
-        Fn(ActionContext<UserType>) -> ActionFutType + Send + Sync + 'static,
-    ActionFutType: Future<Output = ActionFuncResult<ActionFuncOutput, ActionError>>
-        + Send
-        + 'static,
+    for<'c> ActionFuncType: ActionFn<
+        'c,
+        UserType,
+        Output = ActionFuncResult<ActionFuncOutput, ActionError>,
+    >,
     ActionFuncOutput: ActionData,
 {
-    ActionFunc::new_action(f, undo_noop)
+    ActionFunc::new_action(f, |_| async { Ok(()) })
```

Comment:
This might seem silly, but just so I don't lose track of it: can you preserve the `TODO-log` that was was in `undo_noop()`?  I do want to make sure I add a log entry here.

Reply:
Done